### PR TITLE
Feat support concurrency

### PR DESCRIPTION
**Pull request by lachrist**
The initial motivation was to fix https://github.com/getappmap/appmap-agent-js/issues/195 But with jest being kinda distributed, recording actual distributed applications became such a low-hanging fruit that I implemented it.

There is a lot of commits in this PR but I made them small (to my standard anyway) and self-contained (the build should pass at each one of them). There is 4 main parts:

1. `refactor: tweak trace logging` to `refactor: cleanup file indirection`: Remove old sessions which were weak because their only advantage over creating multiple backends was to prevent overwriting appmap files. These old sessions needed to be explicitly open/close. At the beginning of each connection, the first message was expected to be a session.
2. `refactor: add session to some message types` to `refactor: support multiple concurrent sessions in the backend`: Add new session which are associated to one frontend and are included into each message. These sessions mainly enable to differentiate between events that have the same `tab` but comes from different frontends.
3. `refactor: add uuid to process and mocha tracks` to `feat: support recording of concurrent node processes`: Additional changes outside of trace post-processing to make sessions work -- eg: add randomness to track names to prevent collisions and allow tracks to tell which sessions they should be recording
4. `refactor: remove needless spawn component` to `refactor: move logging in spawn`: Enable recorded processes to be spawned outside of the old `batch` component. This is something I had in mind since a long time. It exposed the frontend/backend architecture to the user but I think it makes sense if the recorded application is itself distributed.

### CODE REVIEW

**Comment by brikelly** [comment]
@dividedmind Laurent says you can start your review of this now

**Review by dividedmind — APPROVED** [review]
Looks good, thanks!
Mad props for good commit separation. Makes reviewing even big PRs like this manageable.

File: components/batch/default/index.mjs [hunk 1]
```diff
@@ -69,24 +70,18 @@ export const flushBackendAsync = async (urls, backend, abrupt) => {
...
-    try {
-      return {
-        tokens,
-        ...(await spawnAsync(command, children)),
-      };
-    } catch (error) {
-      logError("Child error %j >> %O", tokens, error);
-      throw new ExternalAppmapError("Failed to spawn child process");
-    }
+  configuration = resolveConfigurationAutomatedRecorder(configuration, env);
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/batch/default/index.mjs]: It is passed to `map` line 120 of this file.

File: components/server/default/index.mjs [hunk 2]
```diff
@@ -119,30 +120,59 @@ export const mainAsync = async (process, configuration) => {
...
+  if (length === 0) {
+    logInfo(
+      "Appmap server listening to client connections at: %j",
+      ports["trace-port"],
+    );
+    logInfo(
+      "Appmap server listening to remote recording requests at: %j",
+      ports["track-port"],
+    );
+    logInfo("Waiting for SIGINT to stop ...");
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/server/default/index.mjs:136]: Thanks!
I disentangle the two commits in
- a35b096646d7b4b89cc12257b5bee527970e3518
- 58eddb122aaa7d10317e9bb2d71bb00e6717362d

**Pull request merged**